api/src/main/java/org/pmiops/workbench/initialcredits/InitialCreditsExpirationServiceImpl.java

```diff
@@ -129,51 +142,58 @@ public void extendInitialCreditsExpiration(DbUser user) {
-          .forEach(
-              ws -> {
-                ws.setInitialCreditsExpired(true);
-                ws.setBillingStatus(BillingStatus.INACTIVE);
-                workspaceDao.save(ws);
-                deleteAppsAndRuntimesInWorkspace(ws);
-              });
-      try {
-        // If the user has already been notified about exhausting their initial credits,
-        // we do not need to notify them about expiration as well.
-        if (!hasExhaustedWorkspace) {
-          mailService.alertUserInitialCreditsExpired(user);
-          userInitialCreditsExpiration.setNotificationStatus(
-              NotificationStatus.EXPIRATION_NOTIFICATION_SENT);
-          userDao.save(user);
-        }
-
-      } catch (MessagingException e) {
-        logger.error(
-            String.format(
-                "Failed to send initial credits expiration notification for user %s",
-                user.getUserId()));
-      }
+    if (null == userInitialCreditsExpiration
+        || userInitialCreditsExpiration.getExpirationCleanupTime() != null) {
+      return;
+    }
+
+    if (haveCreditsExpired(user)) {
+      handleExpiredCredits(user, userInitialCreditsExpiration);
+    } else if (areCreditsExpiringSoon(user)
+        && null == userInitialCreditsExpiration.getApproachingExpirationNotificationTime()) {
+      handleExpiringSoonCredits(user, userInitialCreditsExpiration);
+    }
+  }
+
+  private void handleExpiredCredits(DbUser user, DbUserInitialCreditsExpiration userInitialCreditsExpiration){
+    logger.info(
+        "Initial credits expired for user {}. Expiration time: {}",
+        user.getUsername(),
+        userInitialCreditsExpiration.getExpirationTime());
+
+    workspaceDao.findAllByCreator(user).stream()
+        .filter(
+            ws ->
+                BillingUtils.isInitialCredits(
+                    ws.getBillingAccountName(), workbenchConfigProvider.get()))
+        .filter(DbWorkspace::isActive)
+        .filter(ws -> !ws.isInitialCreditsExpired())
+        .forEach(
+            ws -> {
+              ws.setInitialCreditsExpired(true);
+              ws.setBillingStatus(BillingStatus.INACTIVE);
+              workspaceDao.save(ws);
+              deleteAppsAndRuntimesInWorkspace(ws);
+            });
+
+    userInitialCreditsExpiration.setExpirationCleanupTime(clockNow());
+    userDao.save(user);
+  }
+
+  private void handleExpiringSoonCredits(DbUser user, DbUserInitialCreditsExpiration userInitialCreditsExpiration){
+    logger.info(
+        "Initial credits expiring soon for user {}. Expiration time: {}",
+        user.getUsername(),
+        userInitialCreditsExpiration.getExpirationTime());
+    try {
+      mailService.alertUserInitialCreditsExpiring(user);
+      userInitialCreditsExpiration.setApproachingExpirationNotificationTime(clockNow());
+      userDao.save(user);
+    } catch (MessagingException e) {
+      logger.error(
+          String.format(
```

Comment:
nit: slf4j logger accepts different formatting like what you have in line 184, instead of calling String.format.

Reply:
I was not familiar with that. Thanks for pointing it out. I updated it.